pkg/services/govmomi/util.go

```diff
@@ -66,67 +85,40 @@ func getTask(ctx *context.MachineContext) *mo.Task {
-		switch task.Info.State {
-		case types.TaskInfoStateQueued:
-			ctx.Logger.V(4).Info("task is still pending", "description-id", task.Info.DescriptionId)
-			return true, nil
-		case types.TaskInfoStateRunning:
-			ctx.Logger.V(4).Info("task is still running", "description-id", task.Info.DescriptionId)
-			return true, nil
-		case types.TaskInfoStateSuccess:
-			ctx.Logger.V(4).Info("task is a success", "description-id", task.Info.DescriptionId)
-			ctx.VSphereMachine.Status.TaskRef = ""
-			return false, nil
-		case types.TaskInfoStateError:
-			ctx.Logger.V(2).Info("task failed", "description-id", task.Info.DescriptionId)
-			ctx.VSphereMachine.Status.TaskRef = ""
-			return false, nil
-		default:
-			return false, errors.Errorf("unknown task state %q for %q", task.Info.State, ctx)
-		}
-	}
-
-	return false, nil
-}
+	task := getTask(ctx)
 
-func getMoRef(ctx *context.MachineContext) *types.ManagedObjectReference {
-	if ctx.VSphereMachine.Spec.MachineRef != "" {
-		return &types.ManagedObjectReference{
-			Type:  "VirtualMachine",
-			Value: ctx.VSphereMachine.Spec.MachineRef,
-		}
+	// If no task was found then make sure to clear the VSphereMachine
+	// resource's Status.TaskRef field.
+	if task == nil {
+		ctx.VSphereMachine.Status.TaskRef = ""
+		return false, nil
 	}
 
-	return nil
-}
-
-func getVMfromMachineRef(ctx *context.MachineContext) (*object.VirtualMachine, error) {
-	ref := getMoRef(ctx)
-	if ref == nil {
-		return nil, errors.Errorf("machine ref not set")
+	// Otherwise the course of action is determined by the state of the task.
+	ctx = context.NewMachineLoggerContext(ctx, task.Reference().Value)
+	ctx.Logger.V(4).Info("task found", "state", task.Info.State, "description-id", task.Info.DescriptionId)
+	switch task.Info.State {
+	case types.TaskInfoStateQueued:
+		ctx.Logger.V(4).Info("task is still pending", "description-id", task.Info.DescriptionId)
+		return true, nil
+	case types.TaskInfoStateRunning:
+		ctx.Logger.V(4).Info("task is still running", "description-id", task.Info.DescriptionId)
+		return true, nil
+	case types.TaskInfoStateSuccess:
+		ctx.Logger.V(4).Info("task is a success", "description-id", task.Info.DescriptionId)
+		ctx.VSphereMachine.Status.TaskRef = ""
+		return false, nil
+	case types.TaskInfoStateError:
+		ctx.Logger.V(2).Info("task failed", "description-id", task.Info.DescriptionId)
+		ctx.VSphereMachine.Status.TaskRef = ""
+		return false, nil
+	default:
+		return false, errors.Errorf("unknown task state %q for %q", task.Info.State, ctx)
 	}
-
-	return object.NewVirtualMachine(ctx.Session.Client.Client, *ref), nil
 }
 
-func getVMObject(ctx *context.MachineContext) (mo.VirtualMachine, error) {
-	moRef := types.ManagedObjectReference{
-		Type:  "VirtualMachine",
-		Value: ctx.VSphereMachine.Spec.MachineRef,
-	}
-	var obj mo.VirtualMachine
-	err := ctx.Session.RetrieveOne(ctx, moRef, []string{"name"}, &obj)
-	return obj, err
+func isStringNilOrEmpty(s *string) bool {
```

Comment:
i feel it is slightly easier to read the code if not creating this function, but i am good with both ways

Reply:
Interesting. I wrote it without the helper function the first time, but I thought the function name made it easier to understand. I'll think it over some more. Thanks Hui!

Reply:
Hi @figo,

Good call. I forgot that this function is no longer called in multiple locations. You're right, now it's just overkill.